keystore/filesystem/server_keystore.go

```diff
@@ -387,12 +390,55 @@ func (store *KeyStore) GetPublicKeyFilePath(filename string) string {
 	return fmt.Sprintf("%s%s%s", store.publicKeyDirectory, string(os.PathSeparator), filename)
 }
 
+// use key started with "." (dot) because it's invalid character for clientID that generally stored in cache and
+// it will not intersect with other keys
+const cacheKeySuffix = ".historical."
+
+var errCacheMissHistoricalFilenames = errors.New("cache doesn't contain historical filenames")
+
+func (store *KeyStore) getCachedHistoricalPrivateKeyFilenames(id string) ([]string, error) {
+	key := cacheKeySuffix + id
+	value, ok := store.cache.Get(key)
+	if !ok {
+		return nil, errCacheMissHistoricalFilenames
+	}
+	paths := &fs.HistoricalPaths{}
+	_, err := paths.UnmarshalMsg(value)
+	if err != nil {
+		return nil, err
+	}
+	return paths.Paths, nil
+}
+
+func (store *KeyStore) cacheHistoricalPrivateKeyFilenames(id string, paths []string) error {
+	values := &fs.HistoricalPaths{Paths: paths}
+	serialized, err := values.MarshalMsg(nil)
+	if err != nil {
+		return err
+	}
+	key := cacheKeySuffix + id
+	store.cache.Add(key, serialized)
+	return nil
+}
+
 // GetHistoricalPrivateKeyFilenames return filenames for current and rotated keys
 func (store *KeyStore) GetHistoricalPrivateKeyFilenames(filename string) ([]string, error) {
 	// getHistoricalFilePaths() expects a path, not a name, but we must return names.
 	// Add private key directory path and then remove it to avoid directory switching.
 	fullPath := filepath.Join(store.privateKeyDirectory, filename)
-	paths, err := getHistoricalFilePaths(fullPath, store.fs)
+	paths, err := store.getCachedHistoricalPrivateKeyFilenames(fullPath)
+	if err == nil {
+		return paths, nil
```

Comment:
Guess I'm missing something, still not finished reviewing changes, but how can we be sure that there were no changes in keystore (in filesystem or Redis or whereever) when returning cached (possibly outdated) version?

Reply:
as we discussed previously, for now we guess and it is user's responsibility to take care about it and probably use HTTP API to reset cache after adding new keys. and we have task T2479 to add this feature and track keystore changes